apps/minifront/src/state/ibc-in.ts

```diff
@@ -1,21 +1,209 @@
+
+  const { timeoutHeight, timeoutTimestamp } = await getTimeout(penumbraChainId);
+  const assetMetadata = augmentToAsset(coin.raw.denom, selectedChain.chainName);
+
+  const transferToken = fromDisplayAmount(assetMetadata, coin.displayDenom, amount);
+  const params: MsgTransfer = {
+    sourcePort: 'transfer',
+    sourceChannel: await getCounterpartyChannelId(selectedChain, penumbraChainId),
+    sender: address,
+    receiver: penumbraAddrs.ephemeral,
+    token: transferToken,
+    timeoutHeight,
+    timeoutTimestamp,
+    memo: '',
+  };
+  const ibcTransferMsg = ibc.applications.transfer.v1.MessageComposer.withTypeUrl.transfer(params);
+
+  // TODO: Properly estimate fee using estimateFee() hook from useChain()
+  const estimatedFee: StdFee = {
+    amount: [{ denom: transferToken.denom, amount: '1000' }],
+    gas: '250000',
+  };
+  const client = await getStargateClient();
+  const signedTx = await client.sign(address, [ibcTransferMsg], estimatedFee, '');
+  return await client.broadcastTx(cosmos.tx.v1beta1.TxRaw.encode(signedTx).finish());
+}
+
+const getCounterpartyChannelId = async (
+  counterpartyChain: ChainInfo,
+  penumbraChainId: string,
+): Promise<string> => {
+  const registry = await chainRegistryClient.get(penumbraChainId);
+
+  const counterpartyChannelId = registry.ibcConnections.find(
+    c => c.chainId === counterpartyChain.chainId,
+  )?.counterpartyChannelId;
+  if (!counterpartyChannelId) {
+    throw new Error(
+      `Counterparty channel could not be found in registry for chain id: ${counterpartyChain.chainId}`,
+    );
+  }
+
+  return counterpartyChannelId;
+};
+
+/**
+ * Examples:
+ * getRevisionNumberFromChainId("grand-1") returns 1n
+ * getRevisionNumberFromChainId("osmo-test-5") returns 5n
+ * getRevisionNumberFromChainId("penumbra-testnet-deimos-7") returns 7n
+ */
+export const parseRevisionNumberFromChainId = (chainId: string): bigint => {
+  const match = chainId.match(/-(\d+)$/);
+  if (match?.[1]) {
+    return BigInt(match[1]);
+  } else {
+    throw new Error(`No revision number found within chain id: ${chainId}`);
+  }
+};
+
+// Get timeout from penumbra chain blocks
+const getTimeout = async (chainId: string) => {
+  const { syncInfo } = await tendermintClient.getStatus({});
+  const height = syncInfo?.latestBlockHeight;
+  if (height === undefined) {
+    throw new Error('Could not retrieve latest block height from Tendermint');
+  }
+
+  const timeoutHeight = {
+    revisionNumber: parseRevisionNumberFromChainId(chainId),
+    revisionHeight: height + 1000n, // assuming a block time of 10s and adding ~1000 blocks (~3 hours)
+  };
+  const timeoutTimestamp = currentTimePlusTwoDaysRounded(Date.now());
+
+  return { timeoutHeight, timeoutTimestamp };
+};
+
+const isIbcAsset = (denom: string): boolean => {
+  const ibcRegex = /^ibc\/[0-9A-F]{64}$/i;
+  return ibcRegex.test(denom);
+};
+
+export const isReadySelector = (state: AllSlices) => {
```

Comment:
I'm personally a fan of colocating selectors with the components that use them, unless we expect a given selector to be shared by multiple components. Not a requirement though .